File: c_src/registry.c
#include "../generated/river-input-management.h"
#include "../generated/river-layer-shell.h"
#include "../generated/river-libinput-config.h"
#include "../generated/river-window-management.h"
#include "../generated/river-xkb-bindings.h"
#include "../generated/river-xkb-config.h"
#include <stdio.h>
#include <string.h>
#include <wayland-client-core.h>
#include <wayland-client-protocol.h>

static struct wl_compositor *compositor = NULL;
static struct wl_seat *seat = NULL;
static struct river_window_manager_v1 *river = NULL;
static struct river_xkb_bindings_v1 *xkb_binding = NULL;
static struct river_layer_shell_v1 *layer_shell = NULL;
static struct river_xkb_config_v1 *xkb_config = NULL;
static struct river_input_manager_v1 *input_manager = NULL;
static struct river_libinput_config_v1 *libinput_config = NULL;

static void registry_global(void *data, struct wl_registry *registry,
                            uint32_t name, const char *interface,
                            uint32_t version) {
  // printf("Interface: %s\n", interface);
  if (strcmp(interface, wl_compositor_interface.name) == 0) {
    compositor = wl_registry_bind(registry, name, &wl_compositor_interface, 6);
    printf("Bound wl_compositor!\n");
  }

  if (strcmp(interface, wl_seat_interface.name) == 0) {
    seat = wl_registry_bind(registry, name, &wl_seat_interface, 10);
    printf("Bound wl_seat!\n");
  }

  if (strcmp(interface, river_window_manager_v1_interface.name) == 0) {
    river =
        wl_registry_bind(registry, name, &river_window_manager_v1_interface, 4);
    printf("Bound river_wm!\n");
  }

  if (strcmp(interface, river_xkb_bindings_v1_interface.name) == 0) {
    xkb_binding =
        wl_registry_bind(registry, name, &river_xkb_bindings_v1_interface, 2);
    printf("Bound river_xkb_binding!\n");
  }

  if (strcmp(interface, river_layer_shell_v1_interface.name) == 0) {
    layer_shell =
        wl_registry_bind(registry, name, &river_layer_shell_v1_interface, 1);
    printf("Bound river_layer_shell!\n");
  }

  if (strcmp(interface, river_input_manager_v1_interface.name) == 0) {
    input_manager =
        wl_registry_bind(registry, name, &river_input_manager_v1_interface, 1);
    printf("Bound river_input_management!\n");
  }

  if (strcmp(interface, river_libinput_config_v1_interface.name) == 0) {
    libinput_config = wl_registry_bind(registry, name,
                                       &river_libinput_config_v1_interface, 1);
    printf("Bound river_libinput_config!\n");
  }

  if (strcmp(interface, river_xkb_config_v1_interface.name) == 0) {
    xkb_config =
        wl_registry_bind(registry, name, &river_xkb_config_v1_interface, 1);
    printf("Bound river_xkb_config!\n");
  }
}
static void registry_global_remove(void *data, struct wl_registry *registry,
                                   uint32_t version) {}

static const struct wl_registry_listener registry_listener = {
    .global = registry_global, .global_remove = registry_global_remove};

const struct wl_registry_listener *get_registry_listener(void) {
  return &registry_listener;
}

struct wl_compositor *get_compositor(void) { return compositor; }
struct wl_seat *get_seat(void) { return seat; }
struct river_window_manager_v1 *get_river(void) { return river; }
struct river_xkb_bindings_v1 *get_xkb_bindings(void) { return xkb_binding; }
struct river_layer_shell_v1 *get_layer_shell(void) { return layer_shell; }
struct river_xkb_config_v1 *get_xkb_config(void) { return xkb_config; }
struct river_input_manager_v1 *get_input_manager(void) { return input_manager; }
struct river_libinput_config_v1 *get_libinput_config(void) {
  return libinput_config;
}

File: src/Handlers/Registry.hs
module Handlers.Registry (mkRegistryHandlers) where

import Control.Concurrent.MVar
import Control.Monad (forM_, void)
import Control.Monad.IO.Class (liftIO)
import Data.Map qualified as M
import Data.Set as S (empty)
import Data.Text hiding (show)
import Handlers.InputManagement
import Handlers.LibInput
import Handlers.RiverWM
import Handlers.WlSeat
import Handlers.XkbConfig
import Optics.Core
import Protocols.Generated
import Types
import Wayland.Connection
import Wayland.Generated

mkRegistryHandlers :: MVar WMState -> WlRegistryHandlers
mkRegistryHandlers mvar =
  WlRegistryHandlers
    { onWlRegistryGlobal = bindHandlers mvar
    , onWlRegistryGlobalRemove = removeGlobals mvar
    }

bindHandlers :: MVar WMState -> Object WlRegistry -> Word32 -> Text -> Word32 -> W ()
bindHandlers mvar reg name iface version = case iface of
  "wp_cursor_shape_manager_v1" -> do
    cursor <- wlRegistryBind reg name (min 2 version) WpCursorShapeManagerV1Handlers{}
    modifyMVarW_ mvar $ pure . (#currentCursorShapeManager .~ cursor)
    liftIO $ putStrLn $ "Bound Cursor Shape Manager"
  "river_window_manager_v1" -> do
    wmPtr <- wlRegistryBind reg name (min 5 version) (mkRiverWMHandler mvar)
    modifyMVarW_ mvar $ pure . (#currentWM .~ wmPtr)
    liftIO $ putStrLn $ "Bound Window Manager"
  "river_xkb_bindings_v1" -> do
    xkbBindings <- wlRegistryBind reg name (min 3 version) RiverXkbBindingsV1Handlers{}
    modifyMVarW_ mvar $ pure . (#currentXkbBindings .~ xkbBindings)
    liftIO $ putStrLn $ "Bound Xkb Bindings"
  "river_layer_shell_v1" -> do
    layerShell <- wlRegistryBind reg name (min 1 version) RiverLayerShellV1Handlers{}
    modifyMVarW_ mvar $ pure . (#currentLayerShell .~ layerShell)
    liftIO $ putStrLn $ "Bound Layer Shell"
  "river_input_manager_v1" -> do
    _ <- wlRegistryBind reg name (min 2 version) (mkInputManagerHandlers mvar)
    liftIO $ putStrLn $ "Bound Input Manager"
  "river_libinput_config_v1" -> do
    _ <- wlRegistryBind reg name (min 2 version) (mkLibInputHandlers mvar)
    liftIO $ putStrLn $ "Bound Libinput Config"
  "wl_seat" -> do
    seat <- wlRegistryBind reg name (min 9 version) (mkWlSeatHandlers mvar name)
    modifyMVarW_ mvar $ \state -> do
      let wlSeat =
            WlSeatData
              { wlSeatObj = seat
              , wlSeatCapabilities = S.empty
              , wlPointerSerial = 0
              , wlPointer = Nothing
              , wlCursorShapeDevice = Nothing
              }
      pure $ state & #allWlSeats %~ M.insert name wlSeat
    liftIO $ putStrLn $ "Bound wl_seat: " ++ show name
  "river_xkb_config_v1" -> do
    config <- wlRegistryBind reg name (min 2 version) (mkXkbConfigHandler mvar)
    liftIO $ putStrLn $ "Bound Xkb Config"
    withMVarW mvar $ \WMState{currentKeymapFd} -> do
      forM_ currentKeymapFd $ \fd -> void $ riverXkbConfigV1CreateKeymap config fd RiverXkbConfigV1KeymapFormatTextV1 (mkKeymapHandler mvar)
  _ -> pure ()

removeGlobals :: MVar WMState -> Object WlRegistry -> Word32 -> W ()
removeGlobals mvar _ name = do
  modifyMVarW_ mvar $ \state -> do
    case M.lookup name (state ^. #allWlSeats) of -- Seat Removed
      Nothing -> pure state
      Just wlSeat -> do
        forM_ (wlSeat ^. #wlCursorShapeDevice) wpCursorShapeDeviceV1Destroy
        forM_ (wlSeat ^. #wlPointer) wlPointerRelease
        wlSeatRelease (wlSeat ^. #wlSeatObj)
        pure $ state & #allWlSeats %~ M.delete name
